stlab/concurrency/actor.hpp

```diff
@@ -0,0 +1,179 @@
+/*
+    Copyright 2023 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+/**************************************************************************************************/
+
+#pragma once
+#ifndef STLAB_CONCURRENCY_ACTOR_HPP
+#define STLAB_CONCURRENCY_ACTOR_HPP
+
+// stdc++
+#include <memory>
+#include <optional>
+
+// stlab
+#include <stlab/concurrency/serial_queue.hpp>
+#include <stlab/concurrency/set_current_thread_name.hpp>
+
+// lambda pack captures are a C++20 feature, but both clang and msvc support it, so we should be
+// okay to use it even under C++17.
+#if __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wc++20-extensions"
+#endif // __clang__
+
+/**************************************************************************************************/
+
+namespace stlab {
+inline namespace v1 {
+
+/**************************************************************************************************/
+
+namespace detail {
+
+/**************************************************************************************************/
+
+struct temp_thread_name {
+    explicit temp_thread_name(const char* name) { stlab::set_current_thread_name(name); }
+    ~temp_thread_name() { stlab::set_current_thread_name(""); }
```

Comment:
What if the thread has a name? Isn't this going to clear all the thread names used by the executor?

Reply:
I will add some code here to save off the name and set it back.

Reply:
Done.